### PR TITLE
feat: Implement connection params route

### LINKED ISSUE #8: Implement connection parameters endpoint
`GET /connection-params.der` in the PoHTTP server.

Example:

https://github.com/relaycorp/relaynet-pong/blob/6bc207a06458a3f66cf324cc99d824df29781cd7/src/app/pohttp_endpoint/connectionParams.ts

### PR DESCRIPTION
**Pull request by levansuper**
closes #8

### CODE REVIEW

**Review by gnarea — CHANGES_REQUESTED** [review]
# `Cannot get CryptoKey from secure storage`

I think the error you were getting on Friday was due to the broken ESM support in `@relaycorp/relaynet-core`, which I fixed in https://github.com/relaycorp/relaynet-core-js/pull/561 but didn't get round to testing in this branch. Please test it and report back.

# Vault server missing

Apart from the integration issues I pointed out in `k8s/*.yml`, I noticed that the Vault server was missing from the k8s files. Please save the following under `k8s/vault.yml`:

```yaml
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: vault
spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/min-scale: "1"
    spec:
      containers:
        - name: vault
          image: vault:1.13.2
          command:
            - vault
            - server
            - "-dev"
          env:
            - name: VAULT_DEV_ROOT_TOKEN_ID
              value: "root"
            - name: VAULT_DEV_LISTEN_ADDRESS
              value: "0.0.0.0:$(PORT)"
---
apiVersion: batch/v1
kind: Job
metadata:
  name: vault-config
spec:
  template:
    metadata:
      name: "vault-config"
    spec:
      restartPolicy: OnFailure
      containers:
        - name: vault-config
          image: vault:1.13.2
          imagePullPolicy: IfNotPresent
          command:
            - vault
            - secrets
            - enable
            - "-path=endpoint-keys"
            - kv-v2
          env:
            - name: VAULT_ADDR
              value: "http://vault.default"
            - name: VAULT_TOKEN
              value: root
```

I copied it most of it from the Pong server but didn't have time to test it here, so please try it and report any issues.

File: k8s/pohttp-client.yml [hunk 1]
```diff
@@ -31,6 +31,20 @@ spec:
...
+            - name: INTERNET_ADDRESS
+              value: "example.com"
+            - name: ACTIVE_ID_KEY_REF
+              value: "arn:aws:kms:eu-west-2:111122223333:key/800d5768-3fd7-4edd-a4b8-4c81c3e4c147"
+            - name: KMS_ADAPTER
+              value: "AWS"
+            - name: PRIVATE_KEY_STORE_ADAPTER
+              value: "1"
+            - name: KS_VAULT_URL
+              value: "example.com"
```
[the remark this answers is not in the document]
Comment (gnarea) [reply k8s/pohttp-client.yml:43]: `http://vault.default` if we use the example config I provided.

File: src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts [hunk 2]
```diff
@@ -7,19 +7,22 @@ import type { InternetEndpoint } from '../../awala/InternetEndpoint.js';
...
-  },
-}));
+// jest.unstable_mockModule('../../awala/InternetEndpointManager.js', () => ({
+//   // eslint-disable-next-line @typescript-eslint/naming-convention
+//   InternetEndpointManager: {
+//     init: () => ({
+//       getActiveEndpoint: mockGetActiveEndpoint,
+//     }),
+//   },
+// }));
```
Comment (gnarea) [inline src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts:22]: Remember to undo

File: src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts [hunk 3]
```diff
@@ -28,17 +31,21 @@ describe('fastifyActiveEndpoint', () => {
...
   let activeEndpoint: Partial<InternetEndpoint>;
   beforeEach(() => {
+    mockEndpointManagerInit.mockResolvedValueOnce({
+      getActiveEndpoint: mockGetActiveEndpoint,
+    } as any)
     activeEndpoint = {
       makeInitialSessionKeyIfMissing: mockMakeInitialSessionKey,
     };
-    mockGetActiveEndpoint.mockResolvedValueOnce(activeEndpoint);
+    mockGetActiveEndpoint.mockResolvedValueOnce(activeEndpoint as any);
```
Comment (gnarea) [inline src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts:40]: Why mock this? Unless I made a mistake in the test util for the endpoint manager, this should just work without any mocks or spys.

**Review by gnarea — CHANGES_REQUESTED** [review]

File: src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts (hunk 3 above)
Comment (levansuper) [reply src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts:31]: Tried this way but activeEndpoint is a different object then app.activeEndpoint since getActiveEndpoint always returns new instance
```
  const getConnection = setUpTestDbConnection();
  const getInternetEndpointManager = mockInternetEndpoint(getConnection);
  

  beforeEach(async () => {
    activeEndpoint = await getInternetEndpointManager().getActiveEndpoint();
  })

  test('Fastify should be decorated with active endpoint', async () => {
    const app = fastify();
    await app.register(fastifyMongoose);

    await app.register(fastifyActiveEndpoint);
    await app.ready();

    expect(app).toHaveProperty('activeEndpoint');
    expect(app.activeEndpoint === activeEndpoint).toBeTrue()
  });
  ```

File: src/server/routes/connectionParams.routes.spec.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+  test('A plain simple HEAD request should provide some diagnostic information', async () => {
+    const internetEndpoint = await internetEndpointManager.getActiveEndpoint();
+    const parameters = await internetEndpoint.getConnectionParams();
+
+    const response = await server.inject({ method: 'GET', url: '/connection-params.der' });
+
+    expect(response).toHaveProperty('statusCode', 200);
+    expect(response).toHaveProperty('headers.content-type', 'application/vnd.etsi.tsl.der');
+    expect(response.body).toBe(parameters.toString());
```
[the remark this answers is not in the document]
Comment (levansuper) [reply src/server/routes/connectionParams.routes.spec.ts:22]: Response.body is a string so I just compare it to the "parameters" string value

File: src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts [hunk 5]
```diff
@@ -1,75 +1,76 @@
...
 
-    expect(app).toHaveProperty('activeEndpoint');
-    expect(app.activeEndpoint).toStrictEqual(activeEndpoint);
+    expect(app).toHaveProperty('getActiveEndpoint');
+    const activeEndpoint = await app.getActiveEndpoint();
+    expect(activeEndpoint).toBe(mockActiveEndpoint);
   });
 
-  test('Should make initial session key', async () => {
+  test('Active endpoint should be cached after the first request', async () => {
```
Comment (gnarea) [inline src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts:50]: I think this isn't needed anymore, based on the final verson of the code. It would've been necessary with the previous approach.

**Review by gnarea — CHANGES_REQUESTED** [review]

File: src/server/routes/connectionParams.routes.spec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (levansuper) [reply src/server/routes/connectionParams.routes.spec.ts:21]: This is the log
```
        console.log("Body ===> ", typeof response.body, "parameters ===> ", typeof parameters);

```
```
  Body ===>  string parameters ===>  object
 ```

File: src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    await app.register(fastifyActiveEndpoint);
+    await app.ready();
+
+    expect(app).toHaveProperty('getActiveEndpoint');
+    const activeEndpoint = await app.getActiveEndpoint();
+    expect(activeEndpoint).toBe(mockActiveEndpoint);
+  });
+
+  test('Active endpoint should be created once and cached', async () => {
```
Comment (levansuper) [inline src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts:50]: The title of the test was: 
```
Active endpoint should be created once and cached
```
You suggested to remove that. Maybe we still need a similar test?

NOTE: we are checking that the getActiveEnpoint was called once
`    expect(mockGetActiveEndpoint).toHaveBeenCalledOnce();`

File: src/server/routes/connectionParams.routes.spec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (gnarea) [reply src/server/routes/connectionParams.routes.spec.ts:21]: `response.body` must not be a string. If it's a string, there's a bug in the implementation.

The connection params file is binary -- that's `endpoint.getConnectionParams()` is returning a `Buffer` or `ArrayBuffer`.

File: src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts (hunk 3 above)
Comment (gnarea) [reply src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts]: But why does it matter whether it's the same object? It should be more than enough to know it's equivalent:

```ts
expect(app.activeEndpoint).toMatchObject<InternetEndpoint>(jest.expectObjectContaining({
  id: endpoint.id,
  internetAddress: endpoint.internetAddress,
}));
```

When writing tests, we should be more interested in _outcomes_. Spies and mocks should be used sparingly -- when we can't (easily) inspect the outcomes.

Comment (gnarea) [reply src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts]: Hypothetically, in the future, we could decide to change the following:

```ts
const manager = await InternetEndpointManager.init(conn);
const endpoint = await manager.getActiveEndpoint();
```

to:

```ts
const endpoint = await InternetEndpoint.getActive(conn);
```

If we do that and we keep the tests you wrote originally, we'd have to re-implement a lot of testing code. But why? We're only interested in the outcome.

File: src/server/routes/connectionParams.routes.spec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (gnarea) [reply src/server/routes/connectionParams.routes.spec.ts:21]: Looks like you should be using `response.rawPayload`: https://github.com/relaycorp/relaynet-pong/blob/6bc207a06458a3f66cf324cc99d824df29781cd7/src/app/pohttp_endpoint/connectionParams.spec.ts#L144

Please remember to refer to the code in the Pong repo.

File: src/utilities/fastify/plugins/fastifyActiveEndpoint.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  const endpointManager = await InternetEndpointManager.init(fastify.mongoose);
+  const activeEndpoint = await endpointManager.getActiveEndpoint();
+
+  fastify.decorate('getActiveEndpoint', async () => {
+    await activeEndpoint.makeInitialSessionKeyIfMissing();
+    return activeEndpoint;
+  });
+}
+
+const fastifyActiveEndpointPlugin = fastifyPlugin(fastifyActiveEndpoint, {
```
Comment (gnarea) [inline src/utilities/fastify/plugins/fastifyActiveEndpoint.ts:16]: You still need to specify that we need `fastify.mongoose` somewhere -- we can't lose that information.

See: https://github.com/fastify/fastify-plugin#dependencies. Set either `dependencies` or `decorators`, but not both.

File: src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts (hunk 3 above)
Comment (gnarea) [reply src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts]: One more thing: Even if we -- hypothetically -- wanted to verify that we're calling a method in a particular way, we wouldn't need to mock anything as done here.

We should do something like this:

```ts
// Plan A, if you have access to the instance
const spy = jest.spyOn(instance, 'method');

// Plan B, if you don't have access to the instance
const spy = jest.spyOn(Class.prototype, 'method');
```

... And you'd only do that in the test where you want to check how `Class.method()` is called.

File: src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts (hunk 6 above)
Comment (gnarea) [reply src/utilities/fastify/plugins/fastifyActiveEndpoint.spec.ts:50]: Remember that we're returning a function only as a workaround until #16 is implemented -- this way we can make sure that anything using the active endpoint can use it safely because it's all set up.

We won't need to return a function once #16 is implemented. It'd be a regular attribute, as it was originally: `fastify.activeEndpoint`. So, as an attribute, we wouldn't have to worry about `getActiveEndpoint()` being called multiple times.

TL;DR: This test is correct with the current version of the code, but I'll have to delete pretty soon as part of #16, so we might as well not push it.

**Review by gnarea — CHANGES_REQUESTED** [review]

**Review by gnarea — APPROVED** [review]
Thanks!
